tests/NumericalUnit/Shape.hs

```diff
@@ -17,10 +17,56 @@ import Test.Hspec
         specify "foldr on shape" $  S.foldr (+) 0 (1:* 2:* 3 :* Nil )  `shouldBe`   P.foldr  (+) 0  [1,2,3]
         specify "foldl1 on shape" $  S.foldl1 (+) (1:* 2:* 3 :* Nil )  `shouldBe`   P.foldl1   (+)  [1,2,3]
         specify "foldr1 on shape" $  S.foldr1 (+) (1:* 2:* 3 :* Nil )  `shouldBe`   P.foldr1  (+)  [1,2,3]
 
+        specify "== on Nil" $ Nil == Nil `shouldBe` True
+        specify "== on identical shapes" $ (1:* 2:* 3 :*  Nil) == (1:* 2:* 3 :*  Nil) `shouldBe` True
+        specify "== on different shapes" $ (1:* 2:* 3 :*  Nil) == (1:* 2:* 2 :*  Nil) `shouldBe` False
+
+        specify "fmap on Nil" $ fmap (+1) Nil `shouldBe` Nil
+        specify "fmap on size n shape" $ fmap (+1) (1 :* 3 :* Nil) `shouldBe` (2:* 4 :* Nil)
+
+        specify "pure on Nil" $ pure Nil `shouldBe` Nil
+        specify "pure on size 1" $ (pure 1 :: Shape ('S 'Z) Int) `shouldBe` (1 :* Nil)
+
+        specify "<*> with Nil should always be Nil" $ Nil <*> Nil `shouldBe` Nil
+        specify "<*> with size 1 shape" $ (+) <$> (1 :* Nil) <*> (2 :* Nil) `shouldBe` 3 :* Nil
+
+        specify "traverse on Nil " $ traverse Just Nil `shouldBe` Just Nil
+        specify "traverse on size 1 shape" $ traverse (Just) (1 :* Nil) `shouldBe` Just (1:* Nil)
+        specify "traverse on size n shape" $ traverse (Just) (1 :* 2 :* Nil) `shouldBe` Just (1 :* 2 :* Nil)
+
+        specify "sequenceA on Nil" $ sequenceA Nil `shouldBe` Nil
+        specify "sequenceA on size 1 shape" $ sequenceA (Just 1 :* Nil) `shouldBe` Just (1 :* Nil)
+        specify "sequenceA on size n shape" $ sequenceA (Just 1 :* Just 2 :* Nil) `shouldBe` Just (1 :* 2:*  Nil)
+
+        specify "mapM on Nil" $ mapM (Just) Nil `shouldBe` Just Nil
+        specify "mapM on size 1 shape" $ mapM (Just) (1 :* Nil) `shouldBe` Just (1 :* Nil)
+        specify "mapM on size n shape" $ mapM (Just) (1:* 2 :* Nil) `shouldBe` Just (1:* 2 :* Nil)
+
+        specify "sequence on Nil" $ do a <- sequence Nil ; a `shouldBe` Nil
+        specify "sequence on Size 1 shape" $ sequence (Just 1 :* Nil) `shouldBe` Just (1 :* Nil)
+        specify "sequence on Size n shape" $ sequence (Just 1 :* Just 2 :* Nil) `shouldBe` Just (1 :* 2 :* Nil)
+
+        specify "map on shape" $ S.map (+1) (1:* 2:* 3 :* Nil) `shouldBe` (2:* 3:* 4 :* Nil)
+        specify "map2 on shape" $ S.map2 (+) (1:* 2:* 3 :* Nil) (2:* 3:* 4 :* Nil) `shouldBe` (3:* 5:* 7 :* Nil)
+
+        specify "shapeToList on shape" $ shapeToList (1:* 2:* 3 :*  Nil) `shouldBe` [1, 2, 3]
+        specify "shapeSize on shape" $ shapeSize (1:* 2:* 3 :* Nil)  `shouldBe` length (shapeToList (1:* 2:* 3 :* Nil))
+
+        specify "backwards traverse on shape" $ backwards traverse (:[]) (1:* 2 :* Nil) `shouldBe` [(1:* 2 :* Nil)]
+
+        specify "unShapeVector on shape" $
+          do a <- return (uvFromList [1:* 2:* Nil,2:* 3:* Nil,3:* 4:* Nil :: Shape (S (S Z)) Int])
```

Comment:
heere I think we're testing the Unboxed representation of a sized list (aka shape) right?